File: AuthenticationServices/src/main/java/com/artauction/authenticationservices/Controllers/AuthenticationController.java
package com.artauction.authenticationservices.Controllers;

import com.artauction.authenticationservices.Dtos.Authentication.LoginDto;
import com.artauction.authenticationservices.Dtos.Authentication.LogoutDto;
import com.artauction.authenticationservices.Dtos.Authentication.RegisterArtistDto;
import com.artauction.authenticationservices.Dtos.Authentication.RegisterBuyerDto;
import com.artauction.authenticationservices.Services.Interfaces.AuthenticationServices;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/Authentication")
@RequiredArgsConstructor
public class AuthenticationController {
    // Attributes
    private final AuthenticationServices authenticationService;

    // Buyer Registration API
    @PostMapping("/BuyerRegistration")
    public ResponseEntity<?> registerBuyer(@RequestBody RegisterBuyerDto registerBuyerDto) {
        // Register buyer and check results
        var authenticationDto = authenticationService.registerBuyer(registerBuyerDto);
        if (!authenticationDto.isAuthenticated())
            return ResponseEntity.badRequest().body(authenticationDto.getMessage());

        // Return token
        return ResponseEntity.ok(authenticationDto);
    }

    // Artist Registration API
    @PostMapping("/ArtistRegistration")
    public ResponseEntity<?> registerArtist(@Valid @RequestBody RegisterArtistDto registerArtistDto) {
        // Register artist and check results
        var authenticationDto = authenticationService.registerArtist(registerArtistDto);
        if (!authenticationDto.isAuthenticated())
            return ResponseEntity.badRequest().body(authenticationDto.getMessage());

        // Return token
        return ResponseEntity.ok(authenticationDto);
    }

    // Login API
    @PostMapping("/Login")
    public ResponseEntity<?> login(@Valid @RequestBody LoginDto loginDto) {
        // Login user and check results
        var authenticationDto = authenticationService.login(loginDto);
        if (!authenticationDto.isAuthenticated())
            return ResponseEntity.status(401).body(authenticationDto.getMessage());

        // Return token (with adminId)
        return ResponseEntity.ok(authenticationDto);
    }

    // Refresh token API
    @PostMapping("/RefreshToken")
    public ResponseEntity<?> refreshToken(@RequestBody String refreshToken) {
        // Get refresh token and check
        var authenticationDto = authenticationService.refreshToken(refreshToken);
        if (!authenticationDto.isAuthenticated())
            return ResponseEntity.status(401).body(authenticationDto.getMessage());

        // Return token (with adminId)
        return ResponseEntity.ok(authenticationDto);
    }

    // Logout API
    @PostMapping("/Logout")
    public ResponseEntity<?> logout(@Valid @RequestBody LogoutDto logoutDto) {
        // Check result
        var result = authenticationService.logout(logoutDto.getRefreshToken());
        if (!result)
            return ResponseEntity.badRequest().body("Token is invalid, expired, or already revoked.");

        // Return success status
        return ResponseEntity.ok("Successfully logged out.");
    }
}

File: Project/Backend/src/ArtAuction.API/Controllers/AuthenticationController.cs
    using ArtAuction.Application.DTOs.Authentication;
    using ArtAuction.Application.Interfaces.Services;
    using Microsoft.AspNetCore.Mvc;

    namespace ArtAuction.API.Controllers;

    [ApiController]
    [Route("[controller]")]
    public class AuthenticationController : ControllerBase
    {
        // Attributes
        private readonly IAuthenticationServices _authenticationServices;
        
        // Controller
        public AuthenticationController(IAuthenticationServices authenticationServices)
            =>  _authenticationServices = authenticationServices;
        
        // Buyer registration API
        [HttpPost("BuyerRegistration")]
        public async Task<IActionResult> RegisterBuyer(RegisterBuyerDto registerBuyerDto)
        {
            // Check api state
            if (!ModelState.IsValid)
                return BadRequest(ModelState);
            
            // Register buyer and check results
            var authenticationDto = await _authenticationServices.RegisterBuyer(registerBuyerDto);
            if (!authenticationDto.IsAuthenticated) 
                return BadRequest(authenticationDto.Message);
            
            // Return token
            return Ok( new
            {
                Id = authenticationDto.Id,
                FullName = authenticationDto.Name,
                Email = authenticationDto.Email,
                Role = authenticationDto.Role,
                Token = authenticationDto.Token,
                RefreshToken = authenticationDto.RefreshToken,
                ExpiresOn = authenticationDto.ExpiresOn
            });
        }
        
        // Artist registration API
        [HttpPost("ArtistRegistration")]
        public async Task<IActionResult> RegisterArtist(RegisterArtistDto registerArtistDto)
        {
            // Check api state
            if (!ModelState.IsValid)
                return BadRequest(ModelState);
            
            // Register buyer and check results
            var authenticationDto = await _authenticationServices.RegisterArtist(registerArtistDto);
            if (!authenticationDto.IsAuthenticated) 
                return Unauthorized(authenticationDto.Message);
            
            // Return token
            return Ok( new
            {
                Id = authenticationDto.Id,
                FullName = authenticationDto.Name,
                Email = authenticationDto.Email,
                Role = authenticationDto.Role,
                Token = authenticationDto.Token,
                RefreshToken = authenticationDto.RefreshToken,
                ExpiresOn = authenticationDto.ExpiresOn
            });
        }
        
        // Login API
        [HttpPost("Login")]
        public async Task<IActionResult> Login(LoginDto loginDto)
        {
            // Check api state
            if (!ModelState.IsValid)
                return BadRequest(ModelState);
            
            // Login user and check results
            var authenticationDto = await _authenticationServices.Login(loginDto);
            if (!authenticationDto.IsAuthenticated)
                return Unauthorized(authenticationDto.Message);

            // Return token
            return Ok(new
            {
                Id = authenticationDto.Id,
                FullName = authenticationDto.Name,
                Email = authenticationDto.Email,
                Role = authenticationDto.Role,
                Token = authenticationDto.Token,
                RefreshToken = authenticationDto.RefreshToken,
                ExpiresOn = authenticationDto.ExpiresOn,
                AdminId = authenticationDto.adminId
            });
        }
        
        // Refresh token API
        [HttpPost("RefreshToken")]
        public async Task<IActionResult> RefreshToken([FromBody] string refreshToken)
        {
            // Get refresh token and check 
            var result = await _authenticationServices.RefreshToken(refreshToken);
            if (!result.IsAuthenticated)
                return Unauthorized(result.Message);

            // Return token
            return Ok(new
            {
                Id = result.Id,
                FullName = result.Name,
                Email = result.Email,
                Role = result.Role,
                Token = result.Token,
                RefreshToken = result.RefreshToken,
                ExpiresOn = result.ExpiresOn,
                AdminId = result.adminId
            });
        }
        
        // Logout API
        [HttpPost("Logout")]
        public async Task<IActionResult> Logout([FromBody] LogoutDto logoutDto)
        {
            // Check model state
            if (!ModelState.IsValid)
                return BadRequest(ModelState);

            // Check result
            var result = await _authenticationServices.Logout(logoutDto.refreshToken);
            if (!result)
                return BadRequest("Token is invalid, expired, or already revoked.");

            // Return success stats
            return Ok("Successfully logged out.");
        }
    }